### DISCUSSION TITLE
Request for help understanding / resolving why walls of mesh structures appear as hollow

### DISCUSSION THREAD

**Discussion by topcat** [pyvista/pyvista#5574] [Q&A]
Hello,

First a big thank you to the PyVista team for creating a fantastic and extremely high level library to VTK. 

I am a complete newbie to PyVista, VTK and indeed the entire 3D Modelling / Geometry space and therefore I realize that a big part of the problem I describe below is simply my lack of knowledge. I apologize in advance for this and thank you for your time.

I am using PyVista to view and manipulate mesh files (STL files) in the medical domain. I need to view the STL and then clip it. I used the Box Widget for clipping the STL only to realize that the walls of my object - for example, the wall of the skull, would show as hollow. Since the domain is medical, accuracy is important. 

At first I thought this was a problem after clipping, but when I played with the opacity as I added the mesh with add_mesh(), I realized that the source mesh itself had these internal structures as hollow.

I corroborated that this wasn't a problem in the source STL by creating a simple STL with a cube that has a thick wall, inside the cube is a cylinder also with a thick wall. The image below displays how the mesh appears in various scenarios.

![IMG-20240204-WA0007](https://github.com/pyvista/pyvista/assets/80821/5fc225d1-265f-46db-acdf-bb0ffc0113e3)

The code I am using is simple, essentially :

pl = pv.Plotter()
mesh = pv.read('CR1.stl')
pl.add_mesh(mesh,opacity=0.6)
pl.show()

To further corroborate that this wasn't a problem with the STL files I checked these in software suite by Materialize and in Creo and they show up as fine.

Traversing the discussion thread showed  this discussion https://github.com/pyvista/pyvista/discussions/2649
This seems similar but as I said earlier I don't understand the 3D geometry space and am very new to PyVista to know if this is the same, nor do I understand how to replicate this in my code above.

I would appreciate any help in understanding this issue, pointers to resources and any solutions. 

Thank you so much for your time.

Regards,
Jeetu

**Comment by MatthewFlamm** [comment]
An STL file inherently describes a 2D surface.  In vtk/PyVista terms, this means it is a `PolyData` mesh.  This 2D surface is closed and encloses a volume in your case, but the mesh itself is only the surface.  When you clip it, you end up with a 2D surface that is no longer closed in a 3D sense, unless you add a cap.  Does [`clip_closed_surface`](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.polydatafilters.clip_closed_surface) filter work for this?  I suspect that your Normals must be defined correctly for this to work correctly.

**Comment by topcat** [comment]
Thank you Matthew for taking the time to reply to this.

Your explanation clarifies things quite a bit :)

I was able to apply clip_closed_surface to the mesh for which I had put up
the image in my earlier post and this worked perfectly and I could get the
mesh I wanted :) thanks for this.

Then I went on to use a box widget to cut off a stl of a mesh of a skull.
The box widget returns the clipped region via plotter.box_clipped_meshes
and this is an UnstructuredGrid, so I used extract_surface to convert this
to a Polydata, before applying clip_closed_surface. The code snippet is :

extracted_mesh = pl.box_clipped_meshes[0].extract_surface()
extracted_mesh.clip_closed_surface(normal='-x')
extracted_mesh.save('extracted-mesh.stl')

Unfortunately, this gives me the error :

ValueError: This surface appears to be non-manifold

Am I doing something wrong in the code outlined above or do I need to
somehow repair the mesh before I attempt to use clip_closed_surface()?

Any help would be much appreciated. Thank you for your time.
